optimistic_lock.hpp

```diff
@@ -312,6 +404,25 @@ class [[nodiscard]] optimistic_lock final {
     }
   }
 
+  // Return a read_critical_section for this optimistic_lock using a
+  // version_tag_type which had been obtained previously.  The use case
+  // for this is to fix up the optimistic_lock when a version_tag_type is
+  // read from the stack for an OLC itertor.  It bumps the read lock
+  // count to make the code happy but does not do any spin waits or
+  // even look at the current version_tag_type associated with the lock.
+  // When the caller calls read_critical_section::check() on the
+  // returned lock they will figure out whether or not the version is
+  // still valid.
+  [[nodiscard]] read_critical_section rehydrate_read_lock(
+      version_tag_type version_tag) noexcept {
+    // TODO(laurynas) The inc_read_lock_count call should be
+    // refactored to a RCS-creating factory method in optimistic_lock,
+    // removing the need for this comment and cleaning up usage. Not
+    // necessary to do now.
+    inc_read_lock_count();
```

Comment:
@laurynas-biveinis Please take a look at this line will cause the same problem.  It is used to bump the read_lock_count by the try_prior() and try_next() logic.  In this situation there is no RCS that I can think of on which we could call check() before we take this action.

Reply:
@thompsonbry I don't think that's an issue: in general, I believe `rehydrate_read_lock` has to be written to be similar to `try_read_lock` - except that it has a version arg. In which case `inc_read_lock_count` will only be called if the lock is free at the moment. But even then it will be possible to call `inc_read_lock_count` on RCSs that becomes invalid right after checking that hey were valid - and this can happen currently with `try_read_lock` too.